# Math 134C Week 8

In [2]:
import math
import numpy as np
from scipy.stats import norm
def black_scholes_call(S_t, K, dt, r, sigma, delta):
    d1 = (np.log(S_t / K) + (r -delta + 0.5 * sigma ** 2) * dt) / (sigma * np.sqrt(dt))
    d2 = d1 - sigma * np.sqrt(dt)
    call_price = S_t * np.exp(-delta * dt) * norm.cdf(d1) - K * np.exp(-r * dt) * norm.cdf(d2)
    return call_price

def black_scholes_put(S_t, K, dt, r, sigma, delta):
    d1 = (np.log(S_t / K) + (r -delta + 0.5 * sigma ** 2) * dt) / (sigma * np.sqrt(dt))
    d2 = d1 - sigma * np.sqrt(dt)
    call_price = - S_t * np.exp(-delta * dt) * norm.cdf(-d1) + K * np.exp(-r * dt) * norm.cdf(-d2)
    return call_price

def call_Delta(S_t, K, dt, r, sigma, delta):
    d1 = (np.log(S_t / K) + (r -delta + 0.5 * sigma ** 2) * dt) / (sigma * np.sqrt(dt))  
    return np.exp(-delta * dt) * norm.cdf(d1)

def put_Delta(S_t, K, dt, r, sigma, delta):
    d1 = (np.log(S_t / K) + (r -delta + 0.5 * sigma ** 2) * dt) / (sigma * np.sqrt(dt))  
    return -np.exp(-delta * dt) * norm.cdf(-d1)

def Gamma(S_t, K, dt, r, sigma, delta):
    d1 = (np.log(S_t / K) + (r -delta + 0.5 * sigma ** 2) * dt) / (sigma * np.sqrt(dt))
    return np.exp(-delta * dt) * norm.pdf(d1)/(S_t * sigma * np.sqrt(dt)) 

def call_Theta(S_t, K, dt, r, sigma, delta):
    d1 = (np.log(S_t / K) + (r -delta + 0.5 * sigma ** 2) * dt) / (sigma * np.sqrt(dt))
    d2 = d1 - sigma * np.sqrt(dt)
    return delta * S_t * np.exp(-delta * dt)*norm.cdf(d1) - r * K * np.exp(-r*dt)*norm.cdf(d2)-(K * np.exp(-r*dt) * sigma * norm.pdf(d2))/ (2 * np.sqrt(dt)) 

def put_Theta(S_t, K, dt, r, sigma, delta):
    d1 = (np.log(S_t / K) + (r -delta + 0.5 * sigma ** 2) * dt) / (sigma * np.sqrt(dt))
    d2 = d1 - sigma * np.sqrt(dt)
    return call_Theta(S_t, K, dt, r, sigma, delta) + r * K * np.exp(-r*dt) - delta * S_t * np.exp(-delta * dt)

def put_given_call(C_t, S_t, K, r, dt):
    return C_t - S_t + K * math.exp(-r * dt)
def call_given_put(P_t, S_t, K, r, dt):
    return P_t + S_t - K * math.exp(-r * dt)

**Quiz 7.1**Assume the Black-Scholes framework applies.

Recall that a long 50-strike straddle consists of buying a 50-strike call option and buying a 50-strike put option.

Suppose you enter into a long 50-strike straddle on a nondividend-paying stock and compute the current delta for that straddle with the following information:

- The stock currently sells for 50.
- The stock’s volatility is 50%.
- The time to maturity is one year.
- The continuously compounded risk-free interest rate is 4%.

How much does delta change after three months, if the stock price happens to be 50 at that point in time?

In [6]:
current_delta = call_Delta(50, 50, 1, 0.04, 0.5, 0) + put_Delta(50, 50, 1, 0.04, 0.5, 0)
print(f"The current delta is {current_delta:.6f}")
new_delta = call_Delta(50, 50, 9/12, 0.04, 0.5, 0) + put_Delta(50, 50, 9/12, 0.04, 0.5, 0)
print(f"The new delta is {new_delta:.6f}")
print(f"The difference is {new_delta-current_delta:.6f}")

The current delta is 0.258600
The new delta is 0.224960
The difference is -0.033640


**Quiz 7.2**
Facebook stock has risen in price over 42% so far in 2019.

Consider the "options chain" table of market prices for Facebook (ticker: FB) stock options from May 16, 2019, as reported on the Chicago Board Options Exchange website www.cboe.com.Links to an external site.

The stock price of FB on May 16 was $187.28 and the risk-free annualized interest rate was 2.3%. Please use the column labeled "Last" to find all the market option prices you need in this problem.

![](img/week8_1.png)

The expiration date for all these options is May 31, 2019, so the time to expiration for this problem is 15 days.  Assume the options are European.

Of the 4 strike prices shown in the table, for which strike price is there no put/call parity violation?  

When testing for put/call parity violations, round your numerical computations to the nearest hundredth (round to 2 decimal places).

In [17]:
print("The price of the call for", 182.5, f"strike should be {call_given_put(1.89, 187.28, 182.5, 0.023, 15/365):.2f}")
print("The price of the call for", 185, f"strike should be {call_given_put(2.65, 187.28, 185, 0.023, 15/365):.2f}")
print("The price of the call for", 187.5, f"strike should be {call_given_put(3.85, 187.28, 187.5, 0.023, 15/365):.2f}")
print("The price of the call for", 190, f"strike should be {call_given_put(4.95, 187.28, 190, 0.023, 15/365):.2f}")

The price of the call for 182.5 strike should be 6.84
The price of the call for 185 strike should be 5.10
The price of the call for 187.5 strike should be 3.81
The price of the call for 190 strike should be 2.41


**Quiz 7.3**
Again consider the "options chain" table of market prices for Facebook (ticker: FB) stock options from May 16, 2019, as reported on the Chicago Board Options Exchange website www.cboe.com.  (Links to an external site.)

The stock price of FB on May 16 was 187.28 and the risk-free annualized interest rate was 2.3%. Please use the column labeled "Last" to find all the market option prices you need in this problem.

The expiration date for all these options is May 31, 2019, so the time to expiration for this problem is 15 days. Assume the options are European.

![](img/week8_2.png)

We seek to capture arbitrage profits based on put/call parity violations.  Of the 4 strike prices shown in the table, for which strike price does an arbitrage opportunity exist that involves selling a FB call option and buying a FB put option with the same strike price and collecting an arbitrage profit that is greater than 0.10?

 

In [8]:
print(f"The profit for selling a call option and buying a put option at strike", 182.5, f"is {7.05- call_given_put(1.89, 187.28, 182.5, 0.023, 15/365):.3f}")
print(f"The profit for selling a call option and buying a put option at strike", 185, f"is {4.9- call_given_put(2.65, 187.28, 185, 0.023, 15/365):.3f}")
print(f"The profit for selling a call option and buying a put option at strike", 187.5, f"is {3.35- call_given_put(3.85, 187.28, 187.5, 0.023, 15/365):.3f}")
print(f"The profit for selling a call option and buying a put option at strike", 190, f"is {2.41- call_given_put(4.95, 187.28, 190, 0.023, 15/365):.3f}")

The profit for selling a call option and buying a put option at strike 182.5 is 0.208
The profit for selling a call option and buying a put option at strike 185 is -0.205
The profit for selling a call option and buying a put option at strike 187.5 is -0.457
The profit for selling a call option and buying a put option at strike 190 is 0.000


**Quiz 7.4**
You are given the following information for a stock:

1. The current price of the stock is 50.
2. The stock does not pay dividends.
3. The continuously compounded risk-free interest rate is 8%.
4. A 9-month European call option on the stock has a price of 4.40. The current delta of the call option is 0.5231.

5. A market-maker purchases 1,000 call options and then delta-hedges the position.

A day later, the price of the stock decreased to 45 and the price of the call option decreased to 3.80.

Calculate the one-day profit of the market-maker, rounded to the nearest dollar.

In [14]:
-1000*0.5231*(-5) + 1000*(3.8-4.4) - (-1000 * 0.5231 * 50 + 4.4 * 1000) * (np.exp(0.08/365)-1)


2020.2687417608283

**Quiz 7.5**
You are given the following information for a stock:

1. The current price of the stock is 50.
2. The stock does not pay dividends.
3. The continuously compounded risk-free interest rate is 8%.
4. A 9-month European put option on the stock has a price of 4.40. The current delta of the put option is –0.5701.
5. A market-maker sells 500 put options and then delta-hedges the position.

A day later, the price of the stock increased to 65 and the price of the put option decreased to 3.60.

Calculate the one-day profit of the market-maker, rounded to the nearest dollar.

In [19]:
-500 * (3.6 - 4.4) - 500 * 0.5701 * (65 - 50) - (-500 * 4.4 - 500 * 0.5701 * 50) * (np.exp(0.08/365) - 1)

-3872.1435773927815

**Quiz 7.6**
A market-maker enters into a long position on 100 one-year European straddles with a strike price of $120. A straddle consists of a long call option and a long put option with the same strike price.

The market-maker then delta-hedges it by buying/selling shares of stock and borrowing/lending the remaining balance. You are also given the following assumptions:

1. The value of the stock is 120 now and 125 a day later
2. The continuously compounded risk-free interest rate is 6%
3. The value of a straddle is 46.32 now and 47.86 a day later
4. The current delta of the  straddle is 0.2886

Calculate the one-day profit for the market-maker, rounded to the nearest cent.

In [20]:
100 * (47.86 - 46.32) - 100 * 0.2886 * (125 - 120) - (100 * 46.32 - 100 * 0.2886 * 120) * (np.exp(0.06/365) - 1)

9.507852700654633

**Quiz 7.7**
Assume the Black-Scholes framework for option pricing applies.

A market-maker buys 100 one-year European straddles with the strike price of 200. The market-maker then delta-hedges it by buying/selling shares of a nondividend-paying stock and borrowing/lending the remaining balance. A straddle consists of a long call and a long put on the same strike price. You are given the following information:

1. The value of the stock is 200 now and 185 a day later
2. The stock's volatility is 50%
3. The continuously compounded risk-free interest rate is 5%
4. The cost of one straddle is 77.42 now and 74.10 a day later

5. Calculate the one-day profit for the market-maker, rounded to the nearest cent.

**Quiz 7.8**
For this problem assume

- Stock price S = 33.00
- Volatility $\sigma = 0.32$
- Annual risk-free rate r = 6%
- Continuously paid annual dividend rate $\delta= 0.01$

Consider a European call option on this stock with the following properties:

- Strike price = 35
- Delta = 0.3854
- Gamma = 0.0847
- Days to expiration = 68

Using the delta-gamma approximation, compute the approximate price for the European call option if there is an immediate 0.50 stock price increase.

In [25]:
old_price = black_scholes_call(33, 35, 68/365, 0.06, 0.32, 0.01)
print(f"The original call option price is {black_scholes_call(33, 35, 68/365, 0.06, 0.32, 0.01):.3f}")
new_call_option_price = old_price + 0.3854 * 0.5 + 0.5 * 0.0847 * (0.5) ** 2
print(f"The approximated new price is {new_call_option_price:.2f}")

The original call option price is 1.141
The approximated new price is 1.34


**Quiz 7.9**
Assume the Black-Scholes framework for option pricing applies.

You are given:

- The price of a nondividend-paying stock is 70.
- The continuously compounded risk-free interest rate is 5%.
- The volatility of the stock is greater than 30%.
- A 1-year 72-strike European call option has a price of 10.39. The delta of the option is 0.59381.
- A 1-year 75-strike European call option has a price of 9.16. The delta of the option is 0.54805.

- A market-maker enters into a call bull spread where 100,000 72-strike call options are purchased and 100,000 75-strike call options are sold. The market-maker then delta-hedges the position with shares of the stock.

One day later:

- The price of the stock is 67.
- The price of the 72-strike European call option is 8.67. The delta of the option is 0.54439.
- The price of the 75-strike European call option is 7.57. The delta of the option is 0.49789.

- Determine the number of additional shares of stock that needs to be sold or purchased after one day in order to maintain the delta hedge.

In [38]:
old_delta = 100000 * (0.59381 - 0.54805)
print(f"{old_delta:.2f}")
new_delta = 100000 * (0.54439 - 0.49789)
print(f"{new_delta:.2f}")
print(f"we need to sell {new_delta - old_delta:.2f} shares of the stock to maintain being delta-hedged")

4576.00
4650.00
we need to sell 74.00 shares of the stock to maintain being delta-hedged


**Quiz 7.10** 

Assume the Black-Scholes framework for option pricing applies and the risk-free rate is zero.

AAPL stock is currently trading at 200 while a one-year at-the-money European call option on AAPL is trading at 20, or 10% of the value of the stock.  

What is the implied volatility?

In [33]:
sigma = 0.2515
black_scholes_call(200, 200, 1, 0, sigma, 0)

20.014035660702888

# Exotic Options

## Asian options
- Have payoff based on the average price over some period
- The **average price** option uses the average of the stock price as the asset price when computing the payoff, and the **average strike** option uses the average of the stock price as the strike price.
- The price of an average price options is lower than the corresponding European options. This is because averaging the price reduces the volatility. 

**Example**
Consider the stock price in each month given by $(30, 35, 32, 37, 30, 33)$. Compute the payoff of
1. 6-month arithmetic average price call option at strike 30
2. 6-month arithmetic average price put option at strike 30
3. 6-month geometric average price call option at strike 30
4. 6-month arithmetic average strike put option

In [6]:
arithmetic_average = (30+35+32+37+30+33)/6
geometric_average = (30*35*32*37*30*33)**(1/6)
print(f"The arithmetic average of the price is {arithmetic_average:.2f}")
print(f"The geometric average of the price is {geometric_average:.2f}")
payoff_1 = max(0, arithmetic_average-30)
print(f"1) The payoff is {payoff_1:.2f}")
payoff_2 = max(0, 30-arithmetic_average)
print(f"2) The payoff is {payoff_2:.2f}")
payoff_3 = max(0, geometric_average - 30)
print(f"3) The payoff is {payoff_3:.2f}")
payoff_4 = max(0, arithmetic_average - 33)
print(f"4) The payoff is {payoff_4:.2f}")

The arithmetic average of the price is 32.83
The geometric average of the price is 32.74
1) The payoff is 2.83
2) The payoff is 0.00
3) The payoff is 2.74
4) The payoff is 0.00


**Example**
Consider a three period binomial tree of a stock, where each period is 1 months. Assume that $r = 0.03, \delta = 0.01, u = 1.2, d = 0.8$. Suppose that the current stock price is $40$. Use the binomial tree method to compute the price of a 3-month arithmetic monthly average strike Asian put options. 

**Exercise 14.3** Suppose that S = 100, K = 100, r = 0.08, $\sigma$ = 0.30, $\delta$ = 0, and T = 1. Construct a standard two-period binomial stock price tree using the method in Chapter 10. 

a. Consider stock price averages computed by averaging the 6-month and 1-year prices. What are the possible arithmetic and geometric averages after 1 year? \
b. Construct a binomial tree for the average. How many nodes does it have after 1 year? (Hint: While the moves ud and du give the same year-1 price, they do not give the same average in year 1.) \
c. What is the price of an Asian arithmetic average price call? \
d. What is the price of an Asian geometric average price call?

In [15]:
u = np.exp(0.08 * 0.5 + 0.3 * np.sqrt(0.5))
d = np.exp(0.08 * 0.5 - 0.3 * np.sqrt(0.5))
S = 100
Suu = S * u * u
Sdd = S * d * d
Sud = S * u * d 
q = (np.exp(0.08 * 0.5) - d)/(u-d)

Auu = (S * u + S * u * u)/2 
Aud = (S * u + S * u * d)/2
Adu = (S * d + S * u * d)/2
Add = (S * d + S * d * d)/2

Guu = np.sqrt(S * u * S * u * u)
Gud = np.sqrt(S * u * S * u * d)
Gdu = np.sqrt(S * d * S * d * u)
Gdd = np.sqrt(S * d * S * d * d)

PAu = np.exp(-0.08 * 0.5) * (q * max(Auu-100, 0) + (1-q) * max(Aud - 100, 0))
PAd = np.exp(-0.08 * 0.5) * (q * max(Adu-100, 0) + (1-q) * max(Add-100, 0))
PA = np.exp(-0.08 * 0.5) * (q * (PAu + (1-q) * PAd))

PGu = np.exp(-0.08 * 0.5) * (q * max(Guu-100, 0) + (1-q) * max(Gud - 100, 0))
PGd = np.exp(-0.08 * 0.5) * (q * max(Gdu-100, 0) + (1-q) * max(Gdd-100, 0))
PG = np.exp(-0.08 * 0.5) * (q * (PGu + (1-q) * PGd))
print(f"a) The arithmetic average price after 1 year could be {Auu:.2f}, {Aud:.2f}, {Adu:.2f}, {Add:.2f}")
print(f"   The geometric average price after 1 year could be {Guu:.2f}, {Gud:.2f}, {Gdu:.2f}, {Gdd:.2f}")
print("b) It has 4 notes after 1 year")
print(f"c) The price of the arithematic average price call option is {PA:.2f}")
print(f"d) The price of the geometric average price call option is {PG:.2f}")

a) The arithmetic average price after 1 year could be 147.13, 118.50, 96.26, 77.53
   The geometric average price after 1 year could be 145.97, 118.07, 95.50, 77.24
b) It has 4 notes after 1 year
c) The price of the arithematic average price call option is 12.92
d) The price of the geometric average price call option is 12.61


## Barrier Options
- Have a payoff depending upon whether, over the life of the option, the price of the underlying asset reaches a specified level, called the **barrier**.
- The option came into existence or goes out of existence once the asset price reaches the barrier. If the barrier are not reached, then they can be exercised as ordinary European options.
- The **Knock-out options** go out of existence if the asset price reaches the barrier. If the price has to fall to reach the barrier, the option is **down-and-out**. If the price has to rise to reach the barrier, the option is **up-and-out**
- The **Knock-in options** come into existence if the asset price reaches the barrier. Similarly, we use the terminologies **down-and-in** and **up-and-in**.
- The **Rebate options** make a fixed payment if the asset price reaches the barrier. The payment can occur at the time that the barrier is reacher, or deferred to the expiry date.
- We have "Knock-in" options + "Knock-out" options = ordinary options. This always implies that barrier option hsa price lower than ordinary options, as both of the left hand side term are positive. 

**Exercise 14.6** Let S = 40, K = 45, $\sigma = 0.30$, r = 0.08, T = 1, and $\delta = 0$. 

a. What is the price of a standard call? \
b. What is the price of a knock-in call with a barrier of 44? Why? \
c. What is the price of a knock-out call with a barrier of 44? Why?

In [15]:
price = black_scholes_call(40, 45, 1, 0.08, 0.3, 0)
print(f"a) The price of the standard call is {price:.2f}")
print(f"b) The price of this knock-in call option is the same as the standard call option, which is {price:.2f}")

a) The price of the standard call is 4.13
b) The price of this knock-in call option is the same as the standard call option, which is 4.13


**Example**
Consider a two-year two period binomial model using $S = 100, u = 1.1, d = 0.8, r = 0.05, \delta = 0$. Compute the price of a two-year down-and-in barrier put option with barrier 90 and strike price 120.